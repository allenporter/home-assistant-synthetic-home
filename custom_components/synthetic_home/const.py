"""Constants for Synthetic Home."""

DOMAIN = "synthetic_home"
DEFAULT_NAME = "Synthetic Home"
CONF_FILENAME = "config_filename"

# Used for loading initial state of an entity and it persists across config
# entry reloads. Updated with service calls.
DATA_DEVICE_STATES = "device_states"

ATTR_AREA_NAME = "area"
ATTR_DEVICE_NAME = "device"
ATTR_DEVICE_STATE_KEY = "device_state_key"
ATTR_CONFIG_ENTRY_ID = "config_entry_id"
