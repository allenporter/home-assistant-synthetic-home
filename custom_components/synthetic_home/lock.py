"""Lock platform for Synthetic Home."""

from typing import Any

from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.const import ATTR_CODE
from homeassistant.exceptions import HomeAssistantError
from homeassistant.components.lock import (
    LockEntity,
    LockEntityFeature,
    DOMAIN as LOCK_DOMAIN,
)
from homeassistant.helpers.entity_platform import AddEntitiesCallback

from .const import DOMAIN
from .entity import SyntheticEntity
from .model import ParsedEntity


async def async_setup_entry(
    hass: HomeAssistant, entry: ConfigEntry, async_add_devices: AddEntitiesCallback
) -> None:
    """Set up lock platform."""
    synthetic_home = hass.data[DOMAIN][entry.entry_id]

    async_add_devices(
        SyntheticHomeLock(entity, state=entity.state, **entity.attributes)
        for entity in synthetic_home.entities
        if entity.platform == LOCK_DOMAIN
    )


class SyntheticHomeLock(SyntheticEntity, LockEntity):
    """synthetic_home lock class."""

    _code: str | None = None

    def __init__(
        self,
        entity: ParsedEntity,
        state: str | None = None,
        *,
        supported_features: LockEntityFeature | None = None,
        is_locked: bool | None = None,
        is_locking: bool | None = None,
        is_open: bool | None = None,
        is_opening: bool | None = None,
        is_jammed: bool | None = None,
        code_format: str | None = None,
        code: str | None = None,
    ) -> None:
        """Initialize SyntheticHomeLock."""
        super().__init__(entity)
        if supported_features:
            self._attr_supported_features = LockEntityFeature(0) | supported_features
        if state is not None:
            self._attr_is_locked = state == "locked"
            self._attr_is_locking = state == "locking"
            self._attr_is_open = state == "open"
            self._attr_is_opening = state == "opening"
            self._attr_is_jammed = state == "jammed"
        if is_locked is not None:
            self._attr_is_locked = is_locked
        if is_locking is not None:
            self._attr_is_locking = is_locking
        if is_open is not None:
            self._attr_is_open = is_open
        if is_opening is not None:
            self._attr_is_opening = is_opening
        if is_jammed is not None:
            self._attr_is_jammed = is_jammed
        if code_format is not None:
            self._attr_code_format = code_format
        if code is not None:
            self._code = code

    async def async_lock(self, **kwargs: Any) -> None:
        """Lock the lock.

        This implementation does not require a code to lock.
        """
        # Implementation
        self._attr_is_locked = True
        self.async_write_ha_state()

    async def async_unlock(self, **kwargs: Any) -> None:
        """Unlock the lock.

        This implementation may optionally require a code to unlock.
        """
        if self._code is not None:
            if kwargs.get(ATTR_CODE) != self._code:
                raise HomeAssistantError("Invalid lock code")
        self._attr_is_locked = False
        self.async_write_ha_state()

    async def async_open(self, **kwargs: Any) -> None:
        """Open the door latch."""
        if self._code is not None:
            if kwargs.get(ATTR_CODE) != self._code:
                raise HomeAssistantError("Invalid lock code")
        self._attr_is_open = True
        self._attr_is_locked = False
        self.async_write_ha_state()
