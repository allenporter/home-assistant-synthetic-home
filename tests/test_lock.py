"""Test Synthetic Home lock."""

import pytest
from syrupy import SnapshotAssertion

from homeassistant.const import Platform
from homeassistant.components.lock import (
    SERVICE_LOCK,
    SERVICE_UNLOCK,
    DOMAIN as LOCK_DOMAIN,
)
from homeassistant.const import ATTR_ENTITY_ID
from homeassistant.core import HomeAssistant

from pytest_homeassistant_custom_component.common import MockConfigEntry

from .conftest import HOMES


@pytest.fixture(name="platforms")
def mock_platforms() -> list[Platform]:
    """Set up switch platform."""
    return [Platform.LOCK]


@pytest.mark.parametrize(
    ("config_yaml_fixture", "test_entity"),
    [(f"{HOMES}/smart-lock-example.yaml", "lock.front_door_lock")],
)
async def test_smart_lock(
    hass: HomeAssistant, setup_integration: None, test_entity: str
) -> None:
    """Test a smart lock."""

    state = hass.states.get(test_entity)
    assert state
    assert state.state == "locked"
    assert state.attributes == {
        "friendly_name": "Front Door Lock",
        "supported_features": 0,
    }

    await hass.services.async_call(
        LOCK_DOMAIN,
        SERVICE_UNLOCK,
        service_data={ATTR_ENTITY_ID: test_entity},
        blocking=True,
    )
    await hass.async_block_till_done()
    state = hass.states.get(test_entity)
    assert state
    assert state.state == "unlocked"

    await hass.services.async_call(
        LOCK_DOMAIN,
        SERVICE_LOCK,
        service_data={ATTR_ENTITY_ID: test_entity},
        blocking=True,
    )
    await hass.async_block_till_done()
    state = hass.states.get(test_entity)
    assert state
    assert state.state == "locked"


@pytest.mark.parametrize(
    ("config_yaml_fixture"),
    [
        (f"{HOMES}/smart-lock-locked.yaml"),
        (f"{HOMES}/smart-lock-unlocked.yaml"),
    ],
)
async def test_hvac_device_state(
    hass: HomeAssistant,
    setup_integration: None,
    config_entry: MockConfigEntry,
    snapshot: SnapshotAssertion,
) -> None:
    """Test a lock device with restorable state."""

    state = hass.states.get("lock.front_door_lock")
    assert state
    assert (state.state, state.attributes) == snapshot
