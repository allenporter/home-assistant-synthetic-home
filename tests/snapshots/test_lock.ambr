# serializer version: 1
# name: test_hvac_device_state[tests/fixtures/smart-lock-locked.yaml]
  tuple(
    'locked',
    ReadOnlyDict({
      'friendly_name': 'Front Door Lock',
      'supported_features': <LockEntityFeature: 0>,
    }),
  )
# ---
# name: test_hvac_device_state[tests/fixtures/smart-lock-unlocked.yaml]
  tuple(
    'unlocked',
    ReadOnlyDict({
      'friendly_name': 'Front Door Lock',
      'supported_features': <LockEntityFeature: 0>,
    }),
  )
# ---
